### ISSUE TITLE
Making snappygrav work with data manager plugin with admin

### ISSUE THREAD

**Issue by hughbris** [iusvar/grav-plugin-snappygrav#39]
I've got snappygrav working for user-facing pages just fine. I'm experiencing a few issues getting it to work for submitted data through the data manager plugin.

I've hacked and tried a few options and still have a few more options to explore, but I thought I'd see if you can suggest a simple/elegant solution. Seems it should be simple. Trying each option takes some time, especially since I am preferring to make changes to plugins that can be shared or integrated into plugin repos so that I don't make upgrading difficult for myself.

Ideally I'd like a "PDF download" link/button on the items listing page for each data "type". For now, I would settle for an "Export" button on the full data item page. I've tried simply adding `{{snappgrav(page.route)}}` into the item.html.twig template from the plugin, but I discovered that in the data manager plugin (perhaps all of admin), `page.route` is null. So this can't work unless I find another way to get the page's route.

For the link side, I added a method into the plugin class to simply give me a URL and serve it through snappy-manager. This worked but isn't the URL I wanted, and I get authentication and nonce errors when trying routes within the admin path. I don't quite understand how this plugin routes. Ideally I want the URL to be the same as the data item in data manager but with a .pdf extension.

I can see the [data manager plugin code has something similar for CSV downloads](https://github.com/getgrav/grav-plugin-data-manager/blob/9703833da6016b0a17fd4f05aec00dbac3904df6/data-manager.php#L65) (no docs or anything though) and I should be able to add PDF in the same way. That means hacking two plugins plus creating dependencies and I'm just not sure I (or repo owners) want to go into all that.

Be keen to hear your thoughts on achieving this. This capability would make snappygrav pretty powerful.

**Comment by iusvar** [comment]
I have developed the plugin for users of a site, in fact, even if a few days, snappygrav is earlier.

At this moment I'm more interested in replacing the wkhtmltopdf library that creates too many problems regarding its update and its different versions related to different operating systems.
Actually I think to add two other libraries like tcpdf and mpdf: I do not want to disturb those who are comfortable with wkhtmltopdf.

Until today I had never thought about it, but never ever.

**Comment by hughbris** [comment]
OK thanks, I mostly understand. I don't understand the following, it's just your English phrasing :)

>  in fact, even if a few days, snappygrav is earlier.
 
> .. I had never thought about it, but never ever.

I have managed to achieve what I wanted by only modifying/hacking one plugin, Data Manager. I don't know how to make that into a patch that's useful for everybody though. For now, I will probably maintain a fork and rebase it regularly.

BTW when I figured out how to get the route for an Admin page (`current_route`), I put `{{snappygrav(current_route)|raw}}` in my data item template just to see what happens. I get "Invalid Security Token" when trying to create the PDF, so I think some conflict is happening within the Admin area with tokens and nonces. Possibly because snappy-manager is outside the Admin path. It's OK because I don't need this right now.

If you don't mind me providing some general feedback on your plugin .. it might be more generally useful and extendable if you create smaller methods that do less. For example, `SnappyManager::runWk` could be broken up into methods to:

* determine the PDF creator executable path
* get and set options
* create and return an unencoded PDF, which is more flexible (or implement a method parameter to toggle whether the output is encoded).

And some of the methods, like this one, seem like they should be public, not protected. I just think if you provide more public methods doing smaller jobs, that the plugin will be more flexible and useful to others and perhaps hook in better with other plugins.

Hope that's helpful feedback. It's a great plugin. I would not have known where to start attempting to generate PDFs from form data if it didn't exist. Thanks for your work :)

**Comment by iusvar** [comment]
Thanks for the great advice.

I explain the dark sentences: :-)

> in fact, even if a few days, snappygrav is earlier.

When I wrote this plugin, there was not a data management plugin, so I was not thinking about this possibility.
Even when the data manager was published, I did not think about this, because some libraries already available also provide the output in pdf as dhtmlx.
I'm not saying that a data manager is useless: I do not need it.
Snappygrav was born as a personal need. I did it because it was not expected to be among the priorities of rhuk.

> I had never thought about it, but never ever.

As a result, **yesterday** I did not know if and when I would do it.

However, **today** adopting your advice to create public monofunctional modules I will take into consideration the possibility of doing so.

You wrote:
> Possibly because snappy-manager is outside the Admin path.

Exactly, it works outside the admin environment.

Certainly not enough to include the admin environment to make it work. The changes you suggest will be needed.

Thanks again for your useful feedback.
